leshan-client-core/src/main/java/org/eclipse/leshan/client/resource/LwM2mInstanceEnabler.java

```diff
@@ -97,12 +103,46 @@
      */
     ObserveResponse observe(int resourceid);
 
+    /**
+     * Performs a write of attributes on the instance level, e.g. affecting all resources that haven't specific
+     * attributes themselves.
+     * @return the response object representing the outcome of the operation.
+     */
+    WriteAttributesResponse writeInstanceAttributes(AttributeSet attributes);
+
+    /**
+     * Performs a write of attributes on the resource level.
+     * @param observeSpec The NOTIFICATION class attributes that should be written to the resource
+     * @return the response object representing the outcome of the operation.
+     */
+    WriteAttributesResponse writeResourceAttributes(int resourceId, AttributeSet attributes);
+
+    /**
+     * Performs a Discover on the instance. Link for the instance itself (including attached attributes) and links
+     * for all instantiated resources (including their attributes) must be reported.
+     * @param objectId The object id that this instance belongs to
+     * @param objectAttributes object level attributes, that may contain instance or resource level attributes
+     * @param instanceId The instance id of this instance
+     * @return An array of Link objects, one for the instance and one for each available resource in the instance
+     */
+    Link[] discoverInstance(int objectId, AttributeSet objectAttributes, int instanceId);
```

Comment:
This might feel like it breaks the API by not returning a DiscoverResponse, but I think it's the cleanest way to solve it. When performing discover on the object level, this method will also need to be called and then the links from all instances should be merged into one DiscoverResponse at the object level.